File: frontend/js/views/autopilot.js
/* CollabForge AI — Collab Autopilot (inside Brand Intelligence).
   Guided wizard (no giant dark form) → live 10-stage SSE → campaign pack. */
(function () {
  const CF = window.CF, V = CF.view, fmt = CF.fmt;

  const STEPS = ["Business", "Audience", "Goal", "Budget", "Creator strategy", "Review"];
  const STAGE_LABELS = {
    understand_business: ["Understanding business", "Universal Scraper on the public website"],
    build_campaign_dna: ["Building campaign DNA", "Structuring campaign parameters"],
    discover_creators: ["Discovering creators", "Anakin Search / provided candidates"],
    research_candidates: ["Researching candidates", "Live public evidence per creator"],
    score_and_shortlist: ["Scoring & shortlisting", "Explainable fit score"],
    build_roi_scenarios: ["Building ROI scenarios", "Editable assumptions"],
    draft_outreach: ["Drafting outreach", "Personalized, never auto-sent"],
    draft_campaign_brief: ["Drafting campaign brief", "Grounded in evidence"],
    draft_contract: ["Drafting contract", "Editable template — not legal advice"],
    generate_content_concept: ["Generating content concept", "From real recent titles"],
  };
  const STAGE_ORDER = Object.keys(STAGE_LABELS);

  CF.router.register("brand", "autopilot", async (root) => {
    const biz = CF.state.get("business") || {}; const sel = CF.state.get("selectedCreator");
    const data = {
      business_name: biz.name || "", website: biz.website || "", product: "", industry: "",
      goal: "", audience: "", geography: "India", language: "Hindi/English", budget: "",
      strategy: "auto", chips: sel && sel.name ? [sel.name] : [],
    };
    let step = 0;

    root.innerHTML = V.viewShell(
      V.sectionHead("Workspace", "Collab Autopilot", "From a business brief to a shortlist, outreach, brief and contract — one guided run.") +
      '<div class="card card--pad" id="wiz"></div><div id="ap-run"></div>');
    const wiz = root.querySelector("#wiz");
    const runArea = root.querySelector("#ap-run");

    function stepper() {
      return '<div class="segmented" style="margin-bottom:var(--space-6);flex-wrap:wrap">' +
        STEPS.map((s, i) => '<button class="' + (i === step ? "is-active" : "") + '" data-step="' + i + '">' + (i + 1) + ". " + fmt.escapeHtml(s) + "</button>").join("") + "</div>";
    }
    function field(label, key, ph, tag) {
      tag = tag || "input";
      const val = fmt.inputValue(data[key]);
      return '<div class="field"><label>' + label + "</label>" +
        (tag === "textarea" ? '<textarea class="textarea" data-k="' + key + '" placeholder="' + (ph || "") + '">' + fmt.escapeHtml(val) + "</textarea>"
          : '<input class="input" data-k="' + key + '" value="' + fmt.escapeHtml(val) + '" placeholder="' + (ph || "") + '">') + "</div>";
    }

    function body() {
      if (step === 0) return field("Business name", "business_name", "e.g. NovaPlay Electronics") + field("Website (for Universal Scraper)", "website", "https://…") + field("Product / service", "product", "New gaming smartphone");
      if (step === 1) return field("Target audience", "audience", "Hindi/English tech viewers in India") + '<div class="field-row">' + field("Geography", "geography", "India") + field("Language", "language", "Hindi/English") + "</div>" + field("Industry", "industry", "Consumer electronics");
      if (step === 2) return field("Campaign goal", "goal", "Launch awareness and preorders", "textarea");
      if (step === 3) return '<div class="field-row">' + field("Budget (₹)", "budget", "150000") + "</div><p class=\"hint\">Budget is used for ROI and outreach framing — it is an assumption you can edit.</p>";
      if (step === 4) return creatorStrategy();
      return review();
    }

    function creatorStrategy() {
      return '<div class="field"><label>How should Autopilot choose creators?</label>' +
        '<div class="segmented" id="strat" style="flex-wrap:wrap"><button data-s="auto" class="' + (data.strategy === "auto" ? "is-active" : "") + '">Discover automatically</button>' +
        '<button data-s="shortlist" class="' + (data.strategy === "shortlist" ? "is-active" : "") + '">Use shortlist</button>' +
        '<button data-s="manual" class="' + (data.strategy === "manual" ? "is-active" : "") + '">Enter manually</button></div></div>' +
        '<div class="field"><label>Candidate creators (up to 3)</label><div class="card__meta" id="chips"></div>' +
        '<div class="search-box" style="margin-top:var(--space-2)"><input class="input" id="chip-in" placeholder="Add a creator name and press Enter"><button class="btn btn--ghost" id="chip-add">Add</button></div></div>';
    }

    function review() {
      const rows = [["Business", data.business_name], ["Website", data.website || "—"], ["Product", data.product], ["Goal", data.goal], ["Audience", data.audience], ["Budget", data.budget ? fmt.formatCurrency(+data.budget, "INR") : "—"], ["Creators", data.chips.length ? data.chips.join(", ") : "auto-discover"]];
      return '<div class="table-wrap"><table class="cf-table"><tbody>' + rows.map((r) => '<tr><td class="row-strong">' + r[0] + "</td><td>" + fmt.escapeHtml(r[1] || "—") + "</td></tr>").join("") + "</tbody></table></div>";
    }

    function renderChips() {
      const el = wiz.querySelector("#chips"); if (!el) return;
      el.innerHTML = data.chips.map((c, i) => '<span class="chip chip--removable">' + fmt.escapeHtml(c) + ' <button data-rm="' + i + '">×</button></span>').join("") || '<span class="muted tiny">None yet — auto-discovery will be used.</span>';
      el.querySelectorAll("[data-rm]").forEach((b) => b.addEventListener("click", () => { data.chips.splice(+b.getAttribute("data-rm"), 1); renderChips(); }));
    }

    function render() {
      wiz.innerHTML = stepper() + '<div class="view__section">' + body() + "</div>" +
        '<div style="display:flex;justify-content:space-between;gap:var(--space-2)">' +
        '<button class="btn btn--ghost" id="prev"' + (step === 0 ? " disabled" : "") + ">Back</button>" +
        (step < STEPS.length - 1 ? '<button class="btn btn--primary" id="next">Next</button>' : '<button class="btn btn--primary" id="launch">Launch Autopilot</button>') + "</div>";
      // bind inputs
      wiz.querySelectorAll("[data-k]").forEach((el) => el.addEventListener("input", () => { data[el.getAttribute("data-k")] = el.value; }));
      wiz.querySelectorAll("[data-step]").forEach((b) => b.addEventListener("click", () => { save(); step = +b.getAttribute("data-step"); render(); }));
      const strat = wiz.querySelector("#strat");
      if (strat) strat.querySelectorAll("button").forEach((b) => b.addEventListener("click", () => { data.strategy = b.getAttribute("data-s"); if (data.strategy === "shortlist") data.chips = CF.state.get("shortlistedCreators").map((c) => c.name).slice(0, 3); render(); }));
      const add = wiz.querySelector("#chip-add"), inEl = wiz.querySelector("#chip-in");
      if (add) {
        const addChip = () => { const v = inEl.value.trim(); if (v && data.chips.length < 3) { data.chips.push(v); inEl.value = ""; renderChips(); } };
        add.addEventListener("click", addChip);
        inEl.addEventListener("keydown", (e) => { if (e.key === "Enter") { e.preventDefault(); addChip(); } });
      }
      renderChips();
      const prev = wiz.querySelector("#prev"); if (prev) prev.addEventListener("click", () => { save(); step--; render(); });
      const next = wiz.querySelector("#next"); if (next) next.addEventListener("click", () => { save(); step++; render(); });
      const launch = wiz.querySelector("#launch"); if (launch) launch.addEventListener("click", run);
    }
    function save() { wiz.querySelectorAll("[data-k]").forEach((el) => { data[el.getAttribute("data-k")] = el.value; }); }

    // ── Run: stream the 10-stage autopilot ──
    async function run() {
      if (!data.business_name || !data.product || !data.goal) { CF.ui.toast("Business, product and goal are required", "error"); step = !data.business_name ? 0 : !data.product ? 0 : 2; render(); return; }
      CF.state.set("business", { name: data.business_name, website: data.website, dna: (biz.dna || {}) });
      const stageState = {}; STAGE_ORDER.forEach((s) => stageState[s] = { state: "idle", sub: STAGE_LABELS[s][1], ev: 0, provider: "" });
      let pack = null;
      renderStages(stageState);
      const payload = {
        business_name: data.business_name, website: data.website, product: data.product,
        industry: data.industry, goal: data.goal, audience: data.audience, geography: data.geography,
        language: data.language, budget: data.budget ? +data.budget : null,
        deliverables: "one integrated video segment",
        candidate_creators: data.strategy === "auto" ? [] : data.chips.slice(0, 3),
      };
      try {
        await CF.api.stream("/api/v1/autopilot/run", payload, (ev) => {
          if (ev.stage === "complete") { pack = ev.data || ev.pack || null; return; }
          const st = stageState[ev.stage]; if (!st) return;
          st.state = ev.status === "done" ? "complete" : ev.status === "partial" ? "partial" : ev.status === "failed" ? "failed" : "running";
          if (ev.message) st.sub = ev.message;
          if (ev.provider) st.provider = ev.provider;
          if (ev.evidence_count) st.ev = ev.evidence_count;
          // mark earlier running stages complete
          const idx = STAGE_ORDER.indexOf(ev.stage);
          STAGE_ORDER.forEach((s, i) => { if (i < idx && stageState[s].state === "running") stageState[s].state = "complete"; });
          renderStages(stageState);
        });
      } catch (e) {
        CF.ui.toast("Autopilot stream failed: " + e.message, "error");
      }
      STAGE_ORDER.forEach((s) => { if (stageState[s].state === "running") stageState[s].state = "complete"; });
      renderStages(stageState, pack);
      if (pack) CF.state.set("autopilotRun", { ranAt: Date.now(), selected: pack.selected });
    }

    function renderStages(stageState, pack) {
      let html = '<div class="cf-pipeline view__section" style="margin-top:var(--space-6)"><div class="cf-pipeline__flow">';
      STAGE_ORDER.forEach((s) => {
        const st = stageState[s]; const lab = STAGE_LABELS[s];
        html += '<div class="pl-node" data-state="' + st.state + '"><div class="pl-node__icon">' +
          '<svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2"><circle cx="12" cy="12" r="9"/></svg></div>' +
          '<div class="pl-node__body"><div class="pl-node__title">' + fmt.escapeHtml(lab[0]) +
          '<span class="pl-node__status">' + ({ idle: "Ready", running: "Running", complete: "Done", partial: "Partial", failed: "Failed" }[st.state] || "") + "</span></div>" +
          '<div class="pl-node__sub">' + fmt.escapeHtml(st.sub) + "</div>" +
          (st.provider || st.ev ? '<div class="pl-node__stat">' + (st.provider ? "<span>" + CF.ui.sourceBadge(st.provider) + "</span>" : "") + (st.ev ? "<span>Evidence <b>" + fmt.formatNumber(st.ev) + "</b></span>" : "") + "</div>" : "") +
          "</div></div>";
      });
      html += "</div></div>";
      runArea.innerHTML = html + (pack ? packHtml(pack) : "");
      if (pack) V.wireCopy(runArea);
    }

    function packHtml(pack) {
      const sel = pack.selected || {}; const sl = pack.shortlist || [];
      let h = '<div class="pl-result view__section"><span class="eyebrow">Campaign Launch Pack</span>' +
        '<h3>Top pick: ' + fmt.escapeHtml(sel.name || "—") + "</h3>";
      if (sl.length) h += '<div class="table-wrap view__section"><table class="cf-table"><thead><tr><th>Creator</th><th class="num">Fit</th><th>Confidence</th><th class="num">Reach</th></tr></thead><tbody>' +
        sl.map((c) => '<tr><td class="row-strong">' + fmt.escapeHtml(c.name) + '</td><td class="num">' + fmt.safeText(c.fit, "—") + "</td><td>" + fmt.safeText(c.confidence, "—") + '</td><td class="num">' + fmt.formatCompactNumber(c.subscribers, "—") + "</td></tr>").join("") + "</tbody></table></div>";

      // ROI scenarios
      if (pack.roi && pack.roi.scenarios)
        h += '<div class="view__section"><span class="eyebrow">ROI scenarios</span>' + V.roiTable(pack.roi) + "</div>";

      // Outreach — all drafts, each labelled
      const drafts = (pack.outreach && pack.outreach.drafts) || {};
      const draftEntries = Object.entries(drafts).filter(([, v]) => !fmt.isEmpty(v));
      if (draftEntries.length) {
        h += '<div class="view__section"><span class="eyebrow">Outreach drafts</span>' +
          (pack.outreach.note ? '<p class="tiny muted">' + fmt.escapeHtml(pack.outreach.note) + "</p>" : "") +
          draftEntries.map(([k, v], i) => V.copyBlock("ap-o-" + i, typeof v === "string" ? v : JSON.stringify(v, null, 2))
            .replace("Editable output", fmt.escapeHtml(k.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase())))).join("") + "</div>";
      }

      // Campaign brief
      if (pack.brief) h += '<div class="view__section"><span class="eyebrow">Campaign brief</span>' + V.briefBlock(pack.brief) + "</div>";

      // Content concept
      const cc = pack.content_concept;
      if (cc && (cc.concept_title || cc.hook)) {
        h += '<div class="view__section"><span class="eyebrow">Content concept</span><div class="card card--pad doc-block">';
        if (cc.concept_title) h += "<h4>" + fmt.escapeHtml(cc.concept_title) + "</h4>";
        const cf2 = (label, v) => v ? "<p><strong>" + label + ":</strong> " + fmt.escapeHtml(v) + "</p>" : "";
        h += cf2("Format", cc.format) + cf2("Hook", cc.hook) + cf2("Integration", cc.integration_style) + cf2("CTA", cc.cta) + cf2("Disclosure", cc.disclosure);
        if (Array.isArray(cc.beats) && cc.beats.length) h += "<p><strong>Beats:</strong></p><ul>" + cc.beats.map((b) => "<li>" + fmt.escapeHtml(b) + "</li>").join("") + "</ul>";
        h += "</div></div>";
      }

      // Contract — rendered, not raw markdown
      if (pack.contract && pack.contract.markdown)
        h += '<div class="view__section">' + V.docBlock("ap-contract", pack.contract.markdown, "Contract template") + "</div>";

      if (pack.partial_failures && pack.partial_failures.length) h += CF.ui.warnings(pack.partial_failures.map((p) => ({ message: p })));
      h += '<div style="margin-top:var(--space-4)"><button class="btn btn--primary" id="ap-save">Save to Campaign Room</button></div></div>';
      setTimeout(() => { const b = runArea.querySelector("#ap-save"); if (b) b.addEventListener("click", async () => {
        const res = await CF.api.post("/api/v1/campaigns", { name: (pack.business_dna && pack.business_dna.business_name || data.business_name) + " Autopilot", business_dna: pack.business_dna || {}, shortlist: sl, selected_creator: sel, outreach: pack.outreach, contract: pack.contract, brief: pack.brief, roi: pack.roi, status: "draft" });
        if (res && res.success !== false) CF.ui.toast("Saved to Campaign Room", "success"); else CF.ui.toast("Save failed", "error");
      }); }, 0);
      return h;
    }

    render();
  });
})();

File: backend/api/routes/autopilot.py
"""Collab Autopilot: business brief → creator shortlist → outreach → campaign pack.

Streams 10 stages over SSE. Continues with partial results when a source
fails; failed live data is reported as failed, never swapped for invented data.
"""
import asyncio
import json
import sys, os
from typing import AsyncGenerator, Optional

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from services.security import validate_public_url, research_limiter
from services.generation.business_dna import analyze_business
from services.generation.outreach_generator import generate_outreach
from services.generation.contract_generator import generate_contract
from services.generation.campaign_brief_generator import generate_brief
from services.research.dossier import build_dossier
from services.scoring.roi_scenarios import roi_scenario
from services.anakin.client import anakin_client, AnakinError
from services.ai_providers.openai_client import OpenAIChatClient

router = APIRouter()

STAGES = [
    "understand_business", "build_campaign_dna", "discover_creators",
    "research_candidates", "score_and_shortlist", "build_roi_scenarios",
    "draft_outreach", "draft_campaign_brief", "draft_contract",
    "generate_content_concept",
]


class AutopilotRequest(BaseModel):
    business_name: str = Field(..., min_length=1, max_length=200)
    website: str = Field("", max_length=2000)
    product: str = Field(..., min_length=1, max_length=500)
    industry: str = Field("", max_length=100)
    goal: str = Field(..., min_length=1, max_length=500)
    audience: str = Field("", max_length=300)
    geography: str = Field("", max_length=100)
    language: str = Field("", max_length=60)
    budget: Optional[float] = Field(None, ge=0)
    currency: str = Field("INR", max_length=8)
    deliverables: str = Field("one integrated video segment", max_length=300)
    candidate_creators: list = Field(default_factory=list, max_length=3)
    average_order_value: Optional[float] = Field(None, ge=0)


def _sse(stage: str, status: str, message: str, data: dict | None = None,
         provider: str = "", evidence_count: int = 0, error: str = "") -> str:
    payload = {
        "stage": stage, "status": status, "message": message,
        "provider": provider, "evidence_count": evidence_count,
    }
    if data:
        payload["data"] = data
    if error:
        payload["error"] = error
    return f"data: {json.dumps(payload, ensure_ascii=False, default=str)}\n\n"


async def _run(req: AutopilotRequest) -> AsyncGenerator[str, None]:
    pack: dict = {"partial_failures": []}

    # 1 ── UNDERSTAND BUSINESS
    yield _sse("understand_business", "running", f"Analyzing {req.business_name}…")
    dna = {}
    if req.website:
        try:
            url = validate_public_url(req.website)
            result = await analyze_business(url, f"{req.product} | {req.industry} | {req.goal}")
            if result.get("status") == "ok":
                dna = result["business_dna"]
                yield _sse("understand_business", "done",
                           f"Extracted Business DNA from {url}", {"business_dna": dna},
                           provider="anakin_scrape", evidence_count=len(result.get("evidence", [])))
            else:
                pack["partial_failures"].append("business website analysis failed")
                yield _sse("understand_business", "partial", "Website analysis failed — continuing with your inputs only",
                           error=str(result.get("error", ""))[:200])
        except Exception as e:  # noqa: BLE001 — stage must not kill the stream
            pack["partial_failures"].append("business website analysis failed")
            yield _sse("understand_business", "partial", "Website analysis failed — continuing with your inputs only", error=str(e)[:200])
    else:
        yield _sse("understand_business", "done", "No website provided — using your inputs")
    dna = {**dna, "business_name": req.business_name, "product": req.product,
           "industry": req.industry, "stated_goal": req.goal}
    pack["business_dna"] = dna

    # 2 ── BUILD CAMPAIGN DNA
    yield _sse("build_campaign_dna", "running", "Structuring campaign parameters…")
    campaign = {
        "product": req.product, "industry": req.industry, "goal": req.goal,
        "audience": req.audience, "geography": req.geography, "language": req.language,
        "budget": req.budget, "currency": req.currency, "deliverable": "integrated_segment",
        "deliverables_text": req.deliverables,
    }
    pack["campaign"] = campaign
    yield _sse("build_campaign_dna", "done", "Campaign DNA ready", {"campaign": campaign})

    # 3 ── DISCOVER CREATORS
    yield _sse("discover_creators", "running", "Discovering candidate creators…")
    candidates = [str(c)[:120] for c in req.candidate_creators if str(c).strip()]
    if len(candidates) < 3:
        try:
            res = await anakin_client.search(
                f"List {4 - len(candidates)} popular YouTube creators for {req.audience or req.product} "
                f"in {req.geography or 'their market'} covering {req.product} / {req.industry}. Channel names only."
            )
            names = await _extract_names(res, exclude=candidates)
            candidates.extend(names[: 3 - len(candidates)])
            yield _sse("discover_creators", "done", f"Candidates: {', '.join(candidates)}",
                       {"candidates": candidates}, provider="anakin_search",
                       evidence_count=len(res.get("results", [])))
        except (AnakinError, Exception) as e:  # noqa: BLE001
            if candidates:
                yield _sse("discover_creators", "partial",
                           f"Discovery failed — continuing with provided candidate(s): {', '.join(candidates)}",
                           error=str(e)[:200])
            else:
                yield _sse("discover_creators", "failed", "No candidates available — cannot continue", error=str(e)[:200])
                yield _sse("complete", "failed", "Autopilot stopped: no creators to research", pack)
                return
    else:
        yield _sse("discover_creators", "done", f"Using provided candidates: {', '.join(candidates)}", {"candidates": candidates})

    # 4 ── RESEARCH CANDIDATES
    yield _sse("research_candidates", "running", f"Researching {len(candidates)} creators with live public data…")
    dossiers = []
    for name in candidates[:3]:
        try:
            d = await build_dossier(name, campaign)
            if d.get("status") == "ok":
                dossiers.append(d)
                yield _sse("research_candidates", "progress",
                           f"{d['overview']['name']}: {len(d.get('evidence', []))} evidence items, "
                           f"fit {d['fit_score'].get('score')}",
                           provider=d.get("data_method", ""), evidence_count=len(d.get("evidence", [])))
            else:
                pack["partial_failures"].append(f"research failed for {name}")
                yield _sse("research_candidates", "progress", f"{name}: {d.get('message', 'not found')}", error=d.get("status"))
        except Exception as e:  # noqa: BLE001
            pack["partial_failures"].append(f"research failed for {name}")
            yield _sse("research_candidates", "progress", f"{name}: research error", error=str(e)[:200])
    if not dossiers:
        yield _sse("research_candidates", "failed", "No creator could be researched with real data")
        yield _sse("complete", "failed", "Autopilot stopped: no evidence-backed creators", pack)
        return
    yield _sse("research_candidates", "done", f"Researched {len(dossiers)} creators")

    # 5 ── SCORE AND SHORTLIST
    yield _sse("score_and_shortlist", "running", "Ranking by explainable fit score…")
    # Rank by confidence first, then score; channels without meaningful
    # observable reach can't outrank well-evidenced creators on formula gaps.
    _conf = {"high": 2, "medium": 1, "low": 0}

    def _rank(d):
        fs = d["fit_score"]
        reach_ok = (d["overview"].get("subscribers") or 0) >= 5000
        return (reach_ok, _conf.get(fs.get("confidence"), 0), fs.get("score") or 0)

    dossiers.sort(key=_rank, reverse=True)
    top = dossiers[0]
    shortlist = [{
        "name": d["overview"]["name"], "fit": d["fit_score"].get("score"),
        "confidence": d["fit_score"].get("confidence"),
        "recommendation": d["fit_score"].get("recommendation"),
        "subscribers": d["overview"].get("subscribers"),
        "avg_views": d["engagement_signals"].get("avg_views"),
    } for d in dossiers]
    pack["shortlist"] = shortlist
    pack["selected"] = {"name": top["overview"]["name"], "fit_score": top["fit_score"],
                        "rate_estimate": top.get("rate_estimate")}
    pack["dossiers"] = [{k: d[k] for k in ("overview", "fit_score", "rate_estimate", "engagement_signals", "evidence")} for d in dossiers]
    yield _sse("score_and_shortlist", "done",
               f"Top pick: {top['overview']['name']} (fit {top['fit_score'].get('score')}, "
               f"{top['fit_score'].get('confidence')} confidence)", {"shortlist": shortlist})

    # 6 ── ROI SCENARIOS
    yield _sse("build_roi_scenarios", "running", "Building editable ROI scenario…")
    avg_views = top["engagement_signals"].get("avg_views") or 0
    cost = req.budget or (top.get("rate_estimate") or {}).get("expected") or 0
    if avg_views and cost:
        roi = roi_scenario(
            estimated_impressions=avg_views, engagement_rate_pct=3.0,
            click_through_rate_pct=1.0, conversion_rate_pct=2.0,
            average_order_value=req.average_order_value or 2000.0,
            campaign_cost=cost, currency=req.currency,
        )
        pack["roi"] = roi
        yield _sse("build_roi_scenarios", "done", "ROI scenario ready (all assumptions editable)", {"roi": roi})
    else:
        yield _sse("build_roi_scenarios", "partial", "Insufficient view/cost data for an ROI scenario — skipped")

    # 7 ── OUTREACH
    yield _sse("draft_outreach", "running", "Drafting personalized outreach (never auto-sent)…")
    try:
        outreach = await generate_outreach(
            dna, top["overview"], top.get("recent_content", [])[:2],
            req.goal, req.deliverables,
            f"budget around {req.currency} {req.budget:,.0f}" if req.budget else "request rate card",
            "professional and warm",
        )
        pack["outreach"] = outreach
        yield _sse("draft_outreach", "done", "Outreach drafts ready", {"outreach": outreach})
    except Exception as e:  # noqa: BLE001
        pack["partial_failures"].append("outreach drafting failed")
        yield _sse("draft_outreach", "failed", "Outreach drafting failed", error=str(e)[:200])

    # 8 ── BRIEF
    yield _sse("draft_campaign_brief", "running", "Drafting campaign brief…")
    try:
        brief = await generate_brief(dna, top["overview"], campaign)
        pack["brief"] = brief
        yield _sse("draft_campaign_brief", "done", "Campaign brief ready", {"brief": brief})
    except Exception as e:  # noqa: BLE001
        pack["partial_failures"].append("brief drafting failed")
        yield _sse("draft_campaign_brief", "failed", "Brief drafting failed", error=str(e)[:200])

    # 9 ── CONTRACT
    yield _sse("draft_contract", "running", "Drafting collaboration agreement template…")
    try:
        contract = await generate_contract({
            "brand": req.business_name, "creator": top["overview"]["name"],
            "campaign_name": f"{req.business_name} × {top['overview']['name']}",
            "deliverables": req.deliverables, "platforms": "YouTube",
            "compensation": f"{req.currency} {req.budget:,.0f}" if req.budget else "[TO BE COMPLETED]",
        })
        pack["contract"] = contract
        yield _sse("draft_contract", "done", "Contract template ready (review with counsel)", {"contract": contract})
    except Exception as e:  # noqa: BLE001
        pack["partial_failures"].append("contract drafting failed")
        yield _sse("draft_contract", "failed", "Contract drafting failed", error=str(e)[:200])

    # 10 ── CONTENT CONCEPT
    yield _sse("generate_content_concept", "running", "Generating sponsored content concept…")
    try:
        concept_raw = await OpenAIChatClient().complete(
            "You create sponsored-video concepts grounded in the creator's actual recent content "
            "titles and the campaign brief. No invented statistics. Return ONLY valid JSON.",
            f"Creator: {top['overview']['name']} — recent titles: "
            f"{[v.get('title') for v in top.get('recent_content', [])[:6]]}\n"
            f"Campaign: {json.dumps(campaign)[:800]}\n"
            'Return JSON: {"concept_title": "", "format": "", "hook": "", "beats": ["0-3s ...", ...], '
            '"integration_style": "", "cta": "", "disclosure": ""}',
            max_tokens=600,
        )
        import re as _re
        m = _re.search(r"\{.*\}", concept_raw or "", _re.DOTALL)
        pack["content_concept"] = json.loads(m.group()) if m else {}
        yield _sse("generate_content_concept", "done", "Content concept ready", {"concept": pack["content_concept"]})
    except Exception as e:  # noqa: BLE001
        pack["partial_failures"].append("content concept failed")
        yield _sse("generate_content_concept", "failed", "Content concept failed", error=str(e)[:200])

    yield _sse("complete", "done", "Campaign Launch Pack ready", pack)


@router.post("/run")
async def run_autopilot(req: AutopilotRequest, request: Request):
    research_limiter.check(request)
    return StreamingResponse(
        _run(req), media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


async def _extract_names(search_result: dict, exclude: list) -> list:
    corpus = json.dumps(search_result.get("results", []))[:5000]
    raw = await OpenAIChatClient().complete(
        "Extract YouTube creator/channel names mentioned in this search data. Return ONLY a JSON array of strings.",
        corpus, max_tokens=200,
    )
    import re as _re
    try:
        m = _re.search(r"\[.*\]", raw or "", _re.DOTALL)
        names = json.loads(m.group()) if m else []
    except (json.JSONDecodeError, AttributeError):
        names = []
    ex = {e.lower() for e in exclude}
    return [str(n)[:120] for n in names if isinstance(n, str) and n.lower() not in ex][:4]
